File: frontend/src/app/services/user/user.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {CreateUserRequest} from "./domain/CreateUserRequest";

@Injectable({
	providedIn: 'root'
})
export class UserService
{
	private readonly userUrl: string = '/api/user';

	/**
	 * Will be set to true when user is coming from a url with room id (i.e. joining an already
	 * created room).
	 */
	public onlyJoining: boolean = false;

	constructor(private http: HttpClient)
	{
	}

	/**
	 * Create a user with given username in the room with given roomId
	 * @param username The name of the user to create
	 * @param roomId The id of the room where to create the user
	 */
	public async createUser(username: string, roomId: string): Promise<void>
	{
		const createUserRequest: CreateUserRequest = {
			name: username,
			roomId: roomId
		};
		return this.http.post<void>(this.userUrl, createUserRequest).toPromise();
	}
}

File: backend/src/main/java/de/navvis/pokerplanning/user/UserService.java
package de.navvis.pokerplanning.user;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.stereotype.Service;
import java.util.UUID;
import java.util.regex.Pattern;

import de.navvis.pokerplanning.core.SessionService;
import de.navvis.pokerplanning.room.web.exception.UserAlreadyExistsException;
import de.navvis.pokerplanning.user.web.domain.User;
import de.navvis.pokerplanning.user.web.exception.UnsafeUsernameException;
import de.navvis.pokerplanning.user.web.rest.CreateUserRequest;

@Service
@RequiredArgsConstructor
@Log4j2
public class UserService
{
	private final SessionService sessionService;

	/**
	 * Regex pattern for selecting all unsafe characters to be used for a username. Matches
	 * punctuation: One of !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~.
	 */
	private static final Pattern UNSAFE_CHARS_PATTERN = Pattern.compile("[\\p{Punct}]");

	/**
	 * Create a user by saving his information in the session
	 *
	 * @param userInfo The object that holds user information
	 */
	public User createUser(CreateUserRequest userInfo)
		throws UnsafeUsernameException, UserAlreadyExistsException
	{
		return createUser(userInfo.getName(), userInfo.getRoomId());
	}

	/**
	 * Create a user by saving his information in the session
	 * @param username The name of the user
	 * @param roomId The room where the user is created
	 */
	public User createUser(String username, UUID roomId)
		throws UnsafeUsernameException, UserAlreadyExistsException
	{
		User existingUser = sessionService.getUserForRoom(roomId);
		if (existingUser != null)
		{
			log.error("User already exist in this session. RoomId {}, Existing user {}", roomId,
				existingUser.getName());
			throw new UserAlreadyExistsException();
		}

		validateUsername(username);
		User user = new User(roomId, username);
		sessionService.updateRoomAndUserSessionObject(roomId, user);
		return user;
	}

	/**
	 * Validate the provided username, and throw an exception if it contains unsafe chars
	 * specified in {@link #UNSAFE_CHARS_PATTERN}.
	 * @param username The username to validate
	 * @throws UnsafeUsernameException if unsafe chars are found
	 */
	public void validateUsername(String username) throws UnsafeUsernameException
	{
		if (username == null)
		{
			log.error("The specified username is null, no user will be created");
			throw new UnsafeUsernameException();
		}

		if (UNSAFE_CHARS_PATTERN.matcher(username).find())
		{
			log.error("The specified username '{}' is unsafe, no user will be created", username);
			throw new UnsafeUsernameException();
		}
	}
}
